File: cflib_2_12/java/cflib/src/org.msscf.msscf.cflib.CFLib/org/msscf/msscf/cflib/CFLib/CFLibConsoleMessageLog.java
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright 2020 Mark Stephen Sobkow
 *
 *	This file is part of MSS Code Factory.
 *
 *	MSS Code Factory is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	MSS Code Factory is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with MSS Code Factory.  If not, see https://www.gnu.org/licenses/.
 *
 *	Donations to support MSS Code Factory can be made at
 *	https://www.paypal.com/paypalme2/MarkSobkow
 *
 *	Please contact Mark Stephen Sobkow at msobkow@sasktel.net for commercial licensing.
 */

package org.msscf.msscf.cflib.CFLib;

import java.io.*;
import java.util.*;

/**
 *	Implement the OmfMessageLogInterface over a Log4J Logger.
 */
public class CFLibConsoleMessageLog
implements ICFLibMessageLog {

	/**
	 *	Handle for the current file output stream.
	 */
	private PrintStream	logFile = null;

	/**
	 *	What is the current indent level for logging?
	 */
	private int		   	indent = 0;
	
	/**
	 *	Default constructor.
	 */
	public CFLibConsoleMessageLog() {
		indent = 0;
	}
	
	/**
	 *	Get the current log indent level.
	 */
	public int getMessageLogIndent() {
		return( indent );
	}
	
	/**
	 *	Indent the logging messages another level
	 */
	public synchronized void dedent() {
		if( indent > 0 ) {
			indent --;
		}
		else {
			indent = 0;
		}
	}
	
	/**
	 *	Indent the logging messages another level
	 */
	public synchronized void indent() {
		indent ++;
	}

	/**
	 *	Get an PrintStream that wraps this log
	 */
	public PrintStream getPrintStream() {
		return( logFile );
	}
	
	/**
	 *	Log a message
	 *
	 *	@param		msg		String
	 */
	public synchronized void message( String msg ) {

		if( msg == null ) {
			return;
		}

		StringBuffer buff = new StringBuffer();

		Calendar cal = Calendar.getInstance();
		String stamp = String.format( "%1$tY-%1$tm-%1$td %1$tH:%1$tM:%1$tS.%1$tL ", cal );
		buff.append( stamp );

		int			i;
		for( i = 0; i < indent; i ++ ) {
			buff.append( "\t" );
		}
		buff.append( msg );
		if( ! msg.endsWith( "\n" ) ) {
			buff.append( "\n" );
		}

		boolean wroteSomewhere = false;

		if( logFile != null ) {
			logFile.append( buff.toString() );
			logFile.flush();
		}

		if( ! wroteSomewhere ) {
			System.out.append( buff.toString() );
			System.out.flush();
		}
	}

//	OLD API's that need to go away

	/**
	 *	Open the specified file for logging
	 *
	 *	@param	fileName	The name of the file to open as a log.
	 */
	public void openLogFile( String fileName )
		throws FileNotFoundException
	{
		assert ( fileName != null ) && ( fileName.length() > 0 ) : "File name must be specified";
		assert logFile == null : "Log file is already open";
		logFile = new PrintStream( fileName );
	}
	
	/**
	 *	Close the log file.
	 */
	public void closeLogFile() {
		if( logFile != null ) {
			PrintStream old = logFile;
			logFile = null;
			
			old.flush();
			old.close();
		}
	}

}

File: cflib_2_12/cplus/libcflib/CFLibConsoleMessageLog.cpp
/*
 *  MSS Code Factory CFLib 2.12
 *
 *	Copyright (c) 2020 Mark Stephen Sobkow
 *	
 *	Licensed under the Apache License, Version 2.0 (the "License");
 *	you may not use this file except in compliance with the License.
 *	You may obtain a copy of the License at
 *	
 *	    http://www.apache.org/licenses/LICENSE-2.0
 *	
 *	Unless required by applicable law or agreed to in writing, software
 *	distributed under the License is distributed on an "AS IS" BASIS,
 *	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	See the License for the specific language governing permissions and
 *	limitations under the License.
 */

#include <iostream>
#include <iomanip>
#include <sstream>

#include <cflib/CFLib.hpp>
#include <cflib/ICFLibMessageLog.hpp>
#include <cflib/CFLibConsoleMessageLog.hpp>

namespace cflib {

	CFLibConsoleMessageLog::~CFLibConsoleMessageLog() {
	}

	/**
	 *	Default constructor.
	 */
	CFLibConsoleMessageLog::CFLibConsoleMessageLog() {
		ndnt = 0;
	}

	/**
	 *	Get the current log indent level.
	 */
	int CFLibConsoleMessageLog::getMessageLogIndent() const {
		return( ndnt );
	}

	/**
	 *	Indent the logging messages another level
	 */
	void CFLibConsoleMessageLog::dedent() {
		if( ndnt > 0 ) {
			ndnt --;
		}
		else {
			ndnt = 0;
		}
	}

	/**
	 *	Indent the logging messages another level
	 */
	void CFLibConsoleMessageLog::indent() {
		ndnt ++;
	}

	void CFLibConsoleMessageLog::message( const std::string& msg ) {

		const static std::string S_TAB( "\t" );
		const static std::string S_NEWLINE( "\n" );

		std::string buff;

		time_t systime;
		time( &systime );
		struct tm* tms = gmtime( &systime );
		char fmt[32] = { 0 };
		snprintf( &fmt[0], 32, "%04ld-%02ld-%02ld %02ld:%02ld:%02ld ",
			(long)( tms->tm_year + 1900 ), (long)( tms->tm_mon + 1 ), (long)( tms->tm_mday ),
			(long)(tms->tm_hour), (long)(tms->tm_min), (long)(tms->tm_sec) );
		buff.append( fmt );

		int i;
		for( i = 0; i < ndnt; i ++ ) {
			buff.append( S_TAB );
		}
		buff.append( msg );
		std::string::size_type msglen = msg.length();
		if( msglen > 0 ) {
			if( msg[ msglen - 1 ] != S_NEWLINE[ 0 ] ) {
				buff.append( S_NEWLINE );
			}
		}

		std::cout << buff;
	}

	void CFLibConsoleMessageLog::openLogFile( const std::string& fileName ) {
		// Stub function.  Console wrappers don't write to files.
	}

	void CFLibConsoleMessageLog::closeLogFile() {
		// Stub function.  Console wrappers don't write to files.
	}
}
